ultravox/training/train.py

```diff
@@ -282,48 +297,103 @@ def main() -> None:
         )
-        metrics = evaluation.evaluate(
-            inference,
-            data_dir=args.data_dir,
-            num_procs=args.eval_num_procs,
-            num_samples=args.eval_num_samples,
-            max_new_tokens=args.eval_max_new_tokens,
-            log_dir=wandb.run.dir if wandb.run else str(args.logs_dir),
+        pipeline.save_pretrained(args.output_dir)
+    t_end = datetime.now()
+    logging.info(f"train end time: {t_end}")
+    logging.info(f"elapsed: {t_end - t_start}")
+
+
+def evaluate(args: config_base.TrainConfig):
+    """
+    Evaluate the model on the audio and text datasets.
+
+    NOTE: This function must be run only on the primary process.
+    """
+    logging.info("Starting evaluation...")
+    t_start = datetime.now()
+    logging.info(f"eval start time: {t_start}")
+
+    if args.text_model_lora_config and args.text_model_lora_config.r:
+        logging.warn(
+            "Model has unmerged LoRA config. This can lead to slower inference."
         )
-        if is_master:
-            trainer.log(metrics)
 
-        t_end = datetime.now()
-        logging.info(f"eval end time: {t_end}")
-        logging.info(f"elapsed: {t_end - t_start}")
+    logs_dir = wandb.run.dir if wandb.run else str(args.logs_dir)
+
+    # Run audio-based evaluations and log to W&B
+    audio_metrics = run_oaievalset(
+        log_dir=os.path.join(logs_dir, "oaieval/audio"),
+        model_dir=str(args.output_dir),
+        eval_set="audio-core",
+        num_samples=args.eval_num_samples,
+    )
+    audio_metrics = {f"eval_audio_{k}": v for k, v in audio_metrics.items()}
+    # TODO: it would be best to do trainer.log, but then we'd risk keeping parts of the model
+    # in GPU memory, which could cause OOM errors.
+    if wandb.run:
+        wandb.run.log(audio_metrics)
+
+    # Run text-only evaluations and log to W&B
+    text_metrics = run_oaievalset(
+        log_dir=os.path.join(logs_dir, "oaieval/text"),
+        model_dir=str(args.output_dir),
+        eval_set="transcript-core",
+        num_samples=args.eval_num_samples,
+    )
+    text_metrics = {f"eval_text_{k}": v for k, v in text_metrics.items()}
+    if wandb.run:
+        wandb.run.log(text_metrics)
+
+    t_end = datetime.now()
+    logging.info(f"eval end time: {t_end}")
+    logging.info(f"elapsed: {t_end - t_start}")
+
+
+def run_oaievalset(
+    log_dir: str, model_dir: str, eval_set: str, num_samples: Optional[int] = None
+):
+    env = os.environ.copy()
+
+    # num_gpus = max(1, torch.cuda.device_count())
+    env["EVALS_THREADS"] = "64"
+
+    # TODO: currently running this on a single GPU is faster than multiple GPUs :facepalm:
+    env["CUDA_VISIBLE_DEVICES"] = "0"
+
+    command = [
+        "oaievalset",
+        "--record_dir",
+        log_dir,
+        "generation/gpu/ultravox-dev",
+        eval_set,
+        f"--completion_args=model={model_dir}",
+    ]
+    if num_samples:
+        command.append(f"--max_samples={num_samples}")
```

Comment:
Only doing inference on the first 1024 samples rather than the whole set since evaluation is not as fast as I wanted. If we don't care about the speed, we can do the full set too @zqhuang211.